File: javascript/src/llm/validate.ts
import { Chess } from "chess.js";

import type { ParsedMoveChoice } from "./types.js";

export function resolveLegalMove(
  fen: string,
  choice: ParsedMoveChoice,
  legalUci: string[],
  uciToSan: Map<string, string>,
): { san: string; uci: string } | null {
  const ucisSet = new Set(legalUci);

  if (choice.moveUci) {
    const u = choice.moveUci.toLowerCase();
    if (ucisSet.has(u)) {
      return { san: uciToSan.get(u) ?? "", uci: u };
    }
  }

  if (choice.moveSan) {
    const chess = new Chess(fen);
    const played = chess.move(choice.moveSan.trim());
    if (played) {
      const uci = played.from + played.to + (played.promotion ?? "");
      if (ucisSet.has(uci)) {
        return { san: played.san, uci };
      }
    }
  }
  return null;
}

File: python/src/moltchess/llm/validate.py
from __future__ import annotations

import chess

from .board import legal_moves_from_fen
from .types import ParsedMoveChoice


def resolve_legal_move(
    fen: str,
    choice: ParsedMoveChoice,
    legal_sans: list[str],
    legal_uci: list[str],
    uci_to_san: dict[str, str],
) -> tuple[str, str] | None:
    sans_set = set(legal_sans)
    ucis_set = set(legal_uci)

    if choice.move_uci and choice.move_uci in ucis_set:
        u = choice.move_uci
        return uci_to_san[u], u

    if choice.move_san:
        raw = choice.move_san.strip()
        if raw in sans_set:
            board = chess.Board(fen)
            for m in board.legal_moves:
                if board.san(m) == raw:
                    return raw, m.uci()
        board = chess.Board(fen)
        try:
            move = board.parse_san(raw)
        except ValueError:
            return None
        if move in board.legal_moves:
            return board.san(move), move.uci()
    return None
